quantus/metrics/base.py

```diff
@@ -1,105 +1,331 @@
 
-            # Load a pre-trained LeNet classification model (architecture at quantus/helpers/models).
-            >> model = LeNet()
-            >> model.load_state_dict(torch.load("tutorials/assets/mnist"))
+        # Load a pre-trained LeNet classification model (architecture at quantus/helpers/models).
+        >> model = LeNet()
+        >> model.load_state_dict(torch.load("tutorials/assets/mnist"))
 
-            # Load MNIST datasets and make loaders.
-            >> test_set = torchvision.datasets.MNIST(root='./sample_data', download=True)
-            >> test_loader = torch.utils.data.DataLoader(test_set, batch_size=24)
+        # Load MNIST datasets and make loaders.
+        >> test_set = torchvision.datasets.MNIST(root='./sample_data', download=True)
+        >> test_loader = torch.utils.data.DataLoader(test_set, batch_size=24)
 
-            # Load a batch of inputs and outputs to use for XAI evaluation.
-            >> x_batch, y_batch = iter(test_loader).next()
-            >> x_batch, y_batch = x_batch.cpu().numpy(), y_batch.cpu().numpy()
+        # Load a batch of inputs and outputs to use for XAI evaluation.
+        >> x_batch, y_batch = iter(test_loader).next()
+        >> x_batch, y_batch = x_batch.cpu().numpy(), y_batch.cpu().numpy()
 
-            # Generate Saliency attributions of the test set batch of the test set.
-            >> a_batch_saliency = Saliency(model).attribute(inputs=x_batch, target=y_batch, abs=True).sum(axis=1)
-            >> a_batch_saliency = a_batch_saliency.cpu().numpy()
+        # Generate Saliency attributions of the test set batch of the test set.
+        >> a_batch_saliency = Saliency(model).attribute(inputs=x_batch, target=y_batch, abs=True).sum(axis=1)
+        >> a_batch_saliency = a_batch_saliency.cpu().numpy()
 
-            # Initialise the metric and evaluate explanations by calling the metric instance.
-            >> metric = Metric(abs=True, normalise=False)
-            >> scores = metric(model=model, x_batch=x_batch, y_batch=y_batch, a_batch=a_batch_saliency, **{})
+        # Initialise the metric and evaluate explanations by calling the metric instance.
+        >> metric = Metric(abs=True, normalise=False)
+        >> scores = metric(model=model, x_batch=x_batch, y_batch=y_batch, a_batch=a_batch_saliency}
         """
-        raise NotImplementedError
+        model, x_batch, y_batch, a_batch, s_batch = self.prepare(
+            model=model,
+            x_batch=x_batch,
+            y_batch=y_batch,
+            a_batch=a_batch,
+            s_batch=s_batch,
+            channel_first=channel_first,
+            explain_func=explain_func,
+            explain_func_kwargs=explain_func_kwargs,
+            softmax=softmax,
+            device=device,
+            model_predict_kwargs=model_predict_kwargs,
+            **kwargs,
+        )
+
+        # Create progress bar if desired.
+        iterator = tqdm(
+            enumerate(zip(x_batch, y_batch, a_batch, s_batch)),
```

Comment:
This can be problematic if we need to loop through the _batches more than once.... so we can't set len(x_batch) it also has to account for the # iterations, # runs etc, if that is apart of the metric.

Reply:
I would concur, as you can still implement the multiple runs per batch in the `process_batch` method of each metric (this is implemented for example in the `PixelFlipping` and `RegionPerturbation` metrics).
so as long as the instances themselves are processed independently, this shouldn't be really problematic.
you just need to reorder the loop in your head.

for the progress bar, I don't think that it's necessary to have such a fine-granular resolution. having the number of batches as the denominator in the progress bar should be sufficient. at least for me, I don't see the need to have updates every second as long as I know how long a single iterations takes.
I would otherwise worry to introduce too much metric-specifics in the base-class, as it would then need something like a `nr_runs` parameter, but some metrics will then call this parameter differently and we have chaos again.

but the fix should still be easy if a more granular progressbar is really desired. I would propose to first test lower progressbar-granularity and then only change this if it shows to be really insufficient in practice.